Queue.cpp

```diff
@@ -65,29 +75,51 @@ class Queue::Impl
         }
     }
 
-    std::unique_ptr<Msg> request(Msg&& msg)
+    std::unique_ptr<Msg> request(Msg&& msg, int timeoutMillis)
     {
-        // Construct an ad hoc Queue to handle response Msg
+        Request req;
+
+        // emplace the request in the map
         std::unique_lock<std::mutex> lock(responseMapMutex_);
         auto it = responseMap_.emplace(
-            std::make_pair(msg.getUniqueId(), std::unique_ptr<Queue>(new Queue))).first;
-        lock.unlock();
+            std::make_pair(msg.getUniqueId(), &req)).first;
 
         put(std::move(msg));
-        auto response = it->second->get(); // Block until response is put to the response Queue
 
-        lock.lock();
-        responseMap_.erase(it); // Delete the response Queue
-        lock.unlock();
+        // wait_for returns false if the return is due to timeout
+        if (timeoutMillis <= 0)
+            req._condVar.wait(lock, [&req]{return req._available.load();});
```

Comment:
The response map mutex can't be used here as we are supposed to wait for a response to a one specific request, but that mutex is related to the map holding data for all requests.

Reply:
Holding the map mutex is necesaary as it is needed to remove the iterator in the response map when the request has been successfully proceeded (or not).
That said, as we are touching this map in the request() method, this mutex must be held, because another thread might be pushing a message in the response map.
Otherwise, it would be necessary to create a mutex in the Request class, to wait on it, then additionally hold the map mutex to clean the response state, that might be a big overhead.

Do you agree with this?

Reply:
Actually now that I think about this a bit more this solution seems it could work. I originally thought that this approach would prevent having multiple requests in progress at once, but that doesn't seem to be the case after all as wait()/wait_for() unlock the mutex.

So yes, I do agree! It just seemed unfamiliar to me at first as I had never used multiple condition variables with a single mutex.